src/sources/http_scrape/mod.rs

```diff
@@ -0,0 +1,233 @@
+    )))
+    .take_until(inputs.shutdown)
+    .map(move |_| stream::iter(inputs.urls.clone()))
+    .flatten()
+    .map(move |url| {
+        let client = HttpClient::new(inputs.tls.clone(), &inputs.proxy)
+            .expect("Building HTTP client failed");
+        let endpoint = url.to_string();
+
+        let mut context = context.clone();
+        context.build(&url);
+
+        let mut builder = Request::get(&url).header(http::header::ACCEPT, "text/plain");
+
+        // add user supplied headers
+        if let Some(headers) = &inputs.headers {
+            for header in headers {
+                builder = builder.header(header.0, header.1);
+            }
+        }
+        let mut request = builder.body(Body::empty()).expect("error creating request");
+
+        if let Some(auth) = &inputs.auth {
+            auth.apply(&mut request);
+        }
+
+        let start = Instant::now();
+        client
+            .send(request)
+            .map_err(Error::from)
+            .and_then(|response| async move {
+                let (header, body) = response.into_parts();
+                let body = hyper::body::to_bytes(body).await?;
+                emit!(EndpointBytesReceived {
+                    byte_size: body.len(),
+                    protocol: "http",
+                    endpoint: endpoint.as_str(),
+                });
+                Ok((header, body))
+            })
+            .into_stream()
+            .filter_map(move |response| {
+                ready(match response {
+                    Ok((header, body)) if header.status == hyper::StatusCode::OK => {
+                        emit!(RequestCompleted {
+                            start,
+                            end: Instant::now()
+                        });
+                        match context.on_response(&url, &header, &body) {
+                            Some(events) => {
+                                emit!(HttpScrapeEventsReceived {
+                                    byte_size: events.size_of(),
+                                    count: events.len(),
+                                    uri: url.clone()
+                                });
+
+                                // TODO the below seems wrong placement. It seems should happen
+                                // after the stream is written ? Yet I'm not seeing how to go about
+                                // that.
+                                // emit EventsSent if metrics
+                                if !events.is_empty() {
+                                    if let Event::Metric(ref _metric) =
+                                        events.first().expect("should have event")
+                                    {
+                                        emit!(HttpScrapeEventsSent {
+                                            count: events.len() as u64,
+                                            byte_size: events.size_of()
+                                        });
```

Comment:
I don't think you actually need this one. The associated `component_sent_events_total` metrics are emitted by the topology layer.

Reply:
Ah, I can try removing it 👍

Reply:
And if that's the case, it would answer my question on the TODO above these lines :)

Reply:
Would be interesting if something like this could be detected in a test- like, there are multiple events emited that are duplicates.

Reply:
I wonder if the doc https://github.com/vectordotdev/vector/blob/master/docs/specs/component.md
needs to be updated a bit- there isn't mention of the topology layer at all in that document.

Reply:
That spec is really just meant to cover what a component should do, but doesn't cover how exactly. I think this could be part of internal developer documentation (in particular around adding a new source/sink) though.